File: src/main/java/pl/laskowski/marcin/solving/deduction/combinations/CombinationIterator.java
package pl.laskowski.marcin.solving.deduction.combinations;

import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * Created by Marcin Laskowski.
 */

class CombinationIterator implements Iterator<int[]> {

    private final int collectionSize;
    private int[] current;
    private int[] next;

    CombinationIterator(int sampleSize, int collectionSize) {
        this.current = new int[sampleSize];
        this.next = new int[sampleSize];
        this.collectionSize = collectionSize;
        for (int i = 0; i < next.length; i++) {
            next[i] = i;
        }
    }

    @Override
    public boolean hasNext() {
        return next != null;
    }

    @Override
    public int[] next() {
        if (next == null) throw new NoSuchElementException();
        System.arraycopy(next, 0, current, 0, next.length);
        incrementIndexes();
        return current;
    }

    private void incrementIndexes() {
        next[next.length-1]++;
        int wrongIndex = findFirstTooBigIndex();
        while (wrongIndex > 0) {
            fixIndex(wrongIndex);
            wrongIndex = findFirstTooBigIndex();
        }
        if (wrongIndex == 0) next = null;
    }

    private void fixIndex(int index) {
        next[index - 1]++;
        for (int i = index; i < next.length; i++) {
            int previousValue = next[i - 1];
            next[i] = previousValue + 1;
        }
    }

    private int findFirstTooBigIndex() {
        for (int i = 0; i < next.length; i++) {
            if (next[i] > biggestIndexFor(next, i)) return i;
        }
        return -1;
    }

    private int biggestIndexFor(int[] array, int position) {
        return collectionSize - array.length + position;
    }

}

File: src/main/kotlin/io/github/ilikeyourhat/kudoku/solving/deduction/combinations/CombinationIterator.kt
@file:Suppress("UnsafeCallOnNullableType") // This should be rewritten anyway

package io.github.ilikeyourhat.kudoku.solving.deduction.combinations

internal class CombinationIterator(sampleSize: Int, collectionSize: Int) : MutableIterator<IntArray> {
    private val collectionSize: Int
    private val current: IntArray = IntArray(sampleSize)
    private var next: IntArray?

    init {
        next = IntArray(sampleSize)
        this.collectionSize = collectionSize
        for (i in next!!.indices) {
            next!![i] = i
        }
    }

    override fun hasNext(): Boolean {
        return next != null
    }

    override fun next(): IntArray {
        val next = next ?: throw NoSuchElementException()
        System.arraycopy(next, 0, current, 0, next.size)
        incrementIndexes()
        return current
    }

    override fun remove() {
        TODO("Not yet implemented")
    }

    private fun incrementIndexes() {
        next!![next!!.size - 1]++
        var wrongIndex = findFirstTooBigIndex()
        while (wrongIndex > 0) {
            fixIndex(wrongIndex)
            wrongIndex = findFirstTooBigIndex()
        }
        if (wrongIndex == 0) next = null
    }

    private fun fixIndex(index: Int) {
        next!![index - 1]++
        for (i in index until next!!.size) {
            val previousValue = next!![i - 1]
            next!![i] = previousValue + 1
        }
    }

    private fun findFirstTooBigIndex(): Int {
        for (i in next!!.indices) {
            if (next!![i] > biggestIndexFor(next!!, i)) return i
        }
        return -1
    }

    private fun biggestIndexFor(array: IntArray, position: Int): Int {
        return collectionSize - array.size + position
    }
}
